src/lua_openssl.cc

```diff
@@ -0,0 +1,562 @@
+		}
+		else {
+			SSL_set_connect_state(ssl);
+		}
+	}
+
+	int isHandshakeFinished(SSL* ssl) {
+		return SSL_is_init_finished(ssl);
+	}
+	int doHandshake(SSL* ssl) {
+		return SSL_do_handshake(ssl);
+	}
+
+	int getSSLErrorInfo(const SSL* ssl, const int returnOfSSLOperation) {
+		return SSL_get_error(ssl, returnOfSSLOperation);
+	}
+
+	int readDecriptedDataFromSSL(SSL* ssl, char* buf, const int bufSize) {
+		return SSL_read(ssl, buf, bufSize);
+	}
+
+	int checkDecriptedDataInSSL(SSL* ssl) {
+		char buffer[1];
+		readDecriptedDataFromSSL(ssl, buffer, 0);
+		return SSL_pending(ssl);
+	}
+
+	int writeEncriptedDataIntoSSL(SSL* ssl, const char* buf, const int bufSize) {
+		return SSL_write(ssl, buf, bufSize);
+	}
+
+	void printCertificatesInfoFromSSL(SSL* ssl) {
+    	X509 *cert;
+    	char *line;
+
+    	cert = SSL_get_peer_certificate(ssl);
+    	if ( cert != NULL )
+    	{
+        	printf("Certificates:\n");
+        	line = X509_NAME_oneline(X509_get_subject_name(cert), 0, 0);
+        	printf("Subject: %s\n", line);
+        	free(line);
+        	line = X509_NAME_oneline(X509_get_issuer_name(cert), 0, 0);
+        	printf("Issuer: %s\n", line);
+        	free(line);
+        	X509_free(cert);
+    	}
+    	else
+        printf("No certificates.\n");
+	}
+
+	int checkDataToReadInBIO(BIO* bio) {
+		return BIO_ctrl_pending(bio);
+	}
+
+	int readDatafromBIO(BIO* bio, char* buf, const int bufSize) {
+		return BIO_read(bio, buf, bufSize);
+	}
+
+	int writeDataIntoBIO(BIO* bio, const char* buf, const int bufSize) {
+		return BIO_write(bio, buf, bufSize);
+	}
+
+	// Lua interfaces
+	//OPENSSL
+	int openssl_initSslLib(lua_State* L) {
+		initOpensslLib();
+		return 0;
+	}
+
+	int openssl_cleanupSslLib(lua_State* L) {
+		cleanupOpensslLib();
+		return 0;
+	}
+
+	int openssl_newSslContext(lua_State* L) {
+		const int type = luaL_checknumber(L, 1);
+		SSL_CTX* ctx = newSSLContext(type);
+		if (ctx) {
+			SSL_CTX** pCtx = (SSL_CTX**)lua_newuserdata(L, sizeof(SSL_CTX*));
+			luaL_getmetatable(L, "openssl.ssl_ctx");
+			lua_setmetatable(L, -2);
+			*pCtx = ctx;
+		}
+		else {
+			return luaL_error(L, "Cannot create SSL context with Openssl");
+		}
+		return 1;
+	}
+
+	int openssl_newBio(lua_State* L) {
+		const int type = luaL_checknumber(L, 1);
+		BIO* bio = newBIO(type);
+		if(bio) {
+			BIO** pBio = (BIO**)lua_newuserdata(L, sizeof(BIO*));
+			luaL_getmetatable(L, "openssl.bio");
+			lua_setmetatable(L, -2);
+			*pBio = bio;
+		}
+		else {
+			ERR_print_errors_fp(stderr);
+			return luaL_error(L, "Cannot create BIO with Openssl");
+		}
+		return 1;
+	}
+
+	//SSL_CTX
+	int ctx_initSslContext(lua_State* L) {
+		SSL_CTX* ctx = *(SSL_CTX**)luaL_checkudata(L, 1, "openssl.ssl_ctx");
+		const char* cipherListStr = luaL_checkstring(L, 2);
+		const char* certFile = luaL_checkstring(L, 3);
+		const char* keyFile = luaL_checkstring(L, 4);
+		const int isSuccess = initSSLContext(ctx, cipherListStr, certFile, keyFile);
+		lua_pushboolean(L, isSuccess);
+		return 1;
```

Comment:
@aderiabin this function could be void. Result code is always same

Reply:
@AKalinich-Luxoft This function is a part of `openssl` module interface for Lua.
According C API for Lua each function which can be called from Lua should return `int` value that represent count of results of this function.
So `return 1` mean that function always return 1 value to Lua.